### PR TITLE
os/bluestore: refactor bluefs get_block_extents interface

### PR DESCRIPTION
**Pull request by ifed01**
Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ifed01** [comment]
jenkins test make check

**Comment by benhanokh** [comment]
Igor,
What is the rationale behind this PR?
The only change I can see is that bluefs extents are updated directly into the caller allocator/map(under lock) instead of being copied into a temporary interval-set object and then being pushed to the allocator/map.

How will that solve my problem?
Assuming that the get_bluefs_extenbts logic was sound before it will produce the same results with the changes introduced by this PR.
If the logic was unsound before (and pending changes are still not updated) then we will have the same problem with this PR applied.

The only benefit I can see here is that we save the temporary interval-set object and the double copying of extents first to the interval-set and then to the caller allocator/map.
What did I missed??

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -530,19 +530,20 @@ void BlueFS::dump_block_extents(ostream& out)
...
+  std::function<void(uint64_t, uint32_t)> fn)
 {
   std::lock_guard nl(nodes.lock);
   dout(10) << __func__ << " bdev " << id << dendl;
   ceph_assert(id < alloc.size());
   for (auto& p : nodes.file_map) {
     for (auto& q : p.second->fnode.extents) {
       if (q.bdev == id) {
-        extents->insert(q.offset, q.length);
+        fn(q.offset, q.length);
```
Comment (benhanokh) [inline src/os/bluestore/BlueFS.cc:543]: I don't understand what do we gain by this change.
The logic is exactly like before trusting that nodes.lock guarantees a consistent view of BlueFS allocation with the shared-allocator.
In my testing it seems that there might be some in-flight BlueFS freed-allocations which are falsely showing as being allocated by the shared-allocator and which are *still* reported as being allocated by get_block_extents

**Comment by aclamk** [comment]
I think the real problem is now addressed in #45260
I share @benhanokh reservation on usefulness of this PR.

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ifed01** [comment]
> Igor, What is the rationale behind this PR? The only change I can see is that bluefs extents are updated directly into the caller allocator/map(under lock) instead of being copied into a temporary interval-set object and then being pushed to the allocator/map. How will that help with race condition I'm experiencing with NCB?
> 
> Assuming that the get_bluefs_extents logic was sound before it will produce the same results with the changes introduced by this PR. If the logic was unsound before (and pending changes are still not updated) then we will have the same problem with this PR applied.
> 
> The only benefit I can see here is that we save the temporary interval-set object. What did I missed??

Gabi,
the primary goal of this PR is a code cleanup only. It doesn't relate to any bug fixing including one you're experiencing at the moment.
So indeed the idea is to to avoid all these data tosses in intermediate containers. Which include both interval-set you mentioned and additional vector in load_bluefs_extents.

**Comment by ifed01** [comment]
> I think the real problem is now addressed in #45260 I share @benhanokh reservation on usefulness of this PR.

@aclamk - this is a clean only PR. It fixes no bugfs. See my response to Gabi for more details

**Comment by ifed01** [comment]
jenkins test make check

**Comment by yuriw** [comment]
By @kamoltat 

Failures:

job_id: [6728452, 6728531,  6728608, 6728683]
description: rados/singleton-bluestore/{all/cephtool mon_election/connectivity msgr-failures/none msgr/async objectstore/bluestore-comp-snappy rados supported-random-distro$/{rhel_8}}
failure reason: Command failed (workunit test cephtool/test.sh

2022-03-10T01:40:29.674 INFO:tasks.workunit.client.0.smithi204.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:592: test_auth:  ceph auth get client.xx
2022-03-10T01:40:29.675 INFO:tasks.workunit.client.0.smithi204.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:592: test_auth:  grep caps
2022-03-10T01:40:29.675 INFO:tasks.workunit.client.0.smithi204.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:592: test_auth:  grep mon
2022-03-10T01:40:30.193 INFO:tasks.workunit.client.0.smithi204.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1: test_auth:  rm -fr /tmp/cephtool.gRQ
2022-03-10T01:40:30.194 DEBUG:teuthology.orchestra.run:got remote process result: 1
2022-03-10T01:40:30.196 INFO:tasks.workunit:Stopping ['cephtool'] on client.0...

tracker:


job_id: [6728547]
description: rados/cephadm/workunits/{agent/on mon_election/connectivity task/test_nfs}
failure reason: Test failure: test_cluster_set_reset_user_config (tasks.cephfs.test_nfs.TestNFS)
tracker: https://tracker.ceph.com/issues/48873#change-182596



Dead Jobs:
job_id: [6728462, 6728541, 6728462, 6728619]
description: rados/rook/smoke/{0-distro/ubuntu_20.04 0-kubeadm 0-nvme-loop 1-rook 2-workload/none 3-final cluster/3-node k8s/1.21 net/calico rook/master}
failure reason:  ERROR:tasks.rook:'waiting for service removal' reached maximum tries (90) after waiting for 900 seconds
tracker: https://tracker.ceph.com/issues/53680

**Comment by yuriw** [comment]
this is ready for merge, waiting for `make check` to pass

**Comment by yuriw** [comment]
jenkins test make check

**Comment by yuriw** [comment]
jenkins test make check

**Pull request merged**